File: srcs/requirements/engine/crates/poly-types/src/engine.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum EngineState {
    Idle,
    DiscoveringMarket,
    Ready,
    Quoting,
    HoldingInventory,
    ExpiryGuard,
    RollingOver,
    Reconciling,
}

impl std::fmt::Display for EngineState {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Idle => write!(f, "IDLE"),
            Self::DiscoveringMarket => write!(f, "DISCOVERING_MARKET"),
            Self::Ready => write!(f, "READY"),
            Self::Quoting => write!(f, "QUOTING"),
            Self::HoldingInventory => write!(f, "HOLDING_INVENTORY"),
            Self::ExpiryGuard => write!(f, "EXPIRY_GUARD"),
            Self::RollingOver => write!(f, "ROLLING_OVER"),
            Self::Reconciling => write!(f, "RECONCILING"),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum EngineRunStatus {
    Running,
    Stopped,
    Error,
}

impl std::fmt::Display for EngineRunStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Running => write!(f, "running"),
            Self::Stopped => write!(f, "stopped"),
            Self::Error => write!(f, "error"),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EngineRun {
    pub id: String,
    pub vault_id: String,
    pub status: EngineRunStatus,
    pub current_state: EngineState,
    pub active_market_id: Option<String>,
    pub started_at: i64,
    pub stopped_at: Option<i64>,
    pub last_heartbeat_at: i64,
    pub last_error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CachedBooks {
    pub vault_id: String,
    pub books: serde_json::Value,
    pub updated_at: i64,
}

File: srcs/requirements/frontend/types/engine.ts
// Engine domain types — ported from Python bot/types.py

// ── Enums ──

export type MarketSide = "YES" | "NO";
export type OrderIntent = "BUY" | "SELL";
export type OrderStatus = "OPEN" | "PARTIAL" | "FILLED" | "CANCELLED" | "EXPIRED" | "REJECTED";

// Mirrors the Rust EngineState enum (poly-types). SYNCING_MARKET was previously
// listed here but has no Rust variant, so any UI keyed on it was dead code.
export type EngineState =
  | "IDLE"
  | "DISCOVERING_MARKET"
  | "READY"
  | "QUOTING"
  | "HOLDING_INVENTORY"
  | "EXPIRY_GUARD"
  | "ROLLING_OVER"
  | "RECONCILING"
  | "PAUSED"
  | "ERROR";

export type FillReason =
  | "touched_price"
  | "crossed_top_of_book"
  | "expiry_close"
  | "manual_reconcile"
  | "replay_engine";

// ── Domain Entities ──

export interface ActiveMarket {
  conditionId: string;
  slug: string;
  question: string;
  yesTokenId: string;
  noTokenId: string;
  startTs: number;    // unix seconds
  endTs: number;      // unix seconds
  tickSize: number;
  minOrderSize: number;
}

export interface BookTop {
  bestBid: number | null;
  bestAsk: number | null;
}

export interface VirtualOrder {
  id: string;              // uuid
  engineRunId: string;
  marketId: string;
  side: MarketSide;
  intent: OrderIntent;
  tokenId: string;
  price: number;
  submittedQty: number;
  filledQty: number;
  remainingQty: number;
  status: OrderStatus;
  clientRef: string;       // e.g. "btc5m-buy-yes-1712..-a1b2c3d4"
  placedAt: number;        // unix ms
  expiresAt: number | null;
  simulated: boolean;
  rejectionReason: string | null;
}

export interface SideLedger {
  submittedBuyQty: number;
  filledBuyQty: number;
  openBuyQty: number;
  submittedSellQty: number;
  filledSellQty: number;
  openSellQty: number;
  unsoldInventory: number;
  realizedPnl: number;
  completedCycles: number;
}

export interface MarketState {
  market: ActiveMarket;
  sides: Record<MarketSide, SideLedger>;
  totalNewEntries: number;
  isExpired: boolean;
  buyCancelDone: boolean;
  sellCancelDone: boolean;
  pendingOldMarketId: string | null;
}

export interface EngineRun {
  id: string;
  vaultId: string;
  // Mirrors the Rust EngineRunStatus enum (Running/Stopped/Error). The engine
  // never emits "paused", so any UI matching on it was dead code.
  status: "running" | "stopped" | "error";
  currentState: EngineState;
  activeMarketId: string | null;
  startedAt: number;
  stoppedAt: number | null;
  lastHeartbeatAt: number;
  lastError: string | null;
}

export interface PnlSnapshot {
  vaultId: string;
  runId: string;
  totalRealizedPnl: number;
  totalUnrealizedPnl: number;
  totalPnl: number;
  totalInventoryCost: number;
  totalOpenNotional: number;
  totalCompletedCycles: number;
  sideBreakdown: Record<string, {
    realizedPnl: number;
    unrealizedPnl: number;
    inventoryCost: number;
    openNotional: number;
    unsoldInventory: number;
  }>;
  computedAt: number;
}

export interface AuditRecord {
  type: string;
  timestamp: number;
  [key: string]: string | number | boolean | null | object;
}

// ── Books map: tokenId -> BookTop ──
export type Books = Record<string, BookTop>;

// ── Engine snapshot returned by each cycle ──
export interface EngineSnapshot {
  activeMarket: ActiveMarket | null;
  orders: Record<string, VirtualOrder>;
  marketState: MarketState | null;
  books: Books;
  pnl: PnlSnapshot | null;
}
